src/fraction/mod.rs

```diff
@@ -2337,15 +2336,54 @@ fraction_from_generic_int!(u8, i8, u16, i16, u32, i32, u64, i64, u128, i128, usi
 macro_rules! generic_fraction_from_float {
     ( $($from:ty),*) => {
         $(
-        impl<T: Clone + Integer + CheckedAdd + CheckedMul + CheckedSub> From<$from> for GenericFraction<T> {
-            fn from (val: $from) -> GenericFraction<T> {
-                if val.is_nan () { return GenericFraction::NaN };
-                if val.is_infinite () { return GenericFraction::Infinity (if val.is_sign_negative () { Sign::Minus } else { Sign::Plus }) };
+        impl<T: Clone + FromPrimitive + Integer + CheckedAdd + CheckedMul + CheckedSub> From<$from> for GenericFraction<T> {
+            fn from(val: $from) -> GenericFraction<T> {
+                if val.is_nan () { return Self::NaN };
+                if val.is_infinite () { return Self::Infinity (if val.is_sign_negative () { Sign::Minus } else { Sign::Plus }) };
+
+                let sign = if val < 0.0 { Sign::Minus } else { Sign::Plus };
+
+                // Using https://math.stackexchange.com/a/1049723/17452 , but rely on Ratio::new to compute the gcd.
+                // Find the max precision of this number
+                // Note: the power computations happen in i32 until the end.
+                let mut p: i32 = 0;
+                let mut new_val = val;
+                let ten: $from = 10.0;
+                loop {
+                    if (new_val.floor() - new_val).abs() < <$from>::EPSILON {
+                        // Yay, we've found the precision of this number
+                        break;
+                    }
+                    // Multiply by the precision
+                    // Note: we multiply by powers of ten to avoid this kind of round error with f32s:
+                    // https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=b760579f103b7192c20413ebbe167b90
+                    p += 1;
+                    new_val = val * ten.powi(p);
+                    if new_val.is_infinite() {
+                        // We're overflowing, revert to the string method
+                        let src = format! ("{:+}", val);
```

Comment:
We're using this block of code 3 times later on. To be more DRY I suggest we define a single closure and reuse it in all three places.

I'd consider it to be something like that:
```rust
let fallback_to_string_conversion = || Self::from_decimal_str(&format!("{:+}", val)).unwrap_or(Self::NaN);
```

```rust
return fallback_to_string_conversion()
```